sycl/include/sycl/builtins_utils.hpp

```diff
@@ -0,0 +1,350 @@
+//==-------- builtins_utils.hpp - SYCL built-in function utilities ---------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/detail/boolean.hpp>
+#include <sycl/detail/builtins.hpp>
+#include <sycl/detail/common.hpp>
+#include <sycl/detail/generic_type_traits.hpp>
+#include <sycl/detail/iostream_proxy.hpp>
+#include <sycl/pointers.hpp>
+#include <sycl/types.hpp>
+
+#include <algorithm>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+
+#ifdef __SYCL_DEVICE_ONLY__
+#define __sycl_std
+#else
+namespace __sycl_std = __host_std;
+#endif
+
+namespace detail {
+
+// Get the element type of T. If T is a scalar, the element type is considered
+// the type of the scalar.
+template <typename T> struct get_elem_type {
+  using type = T;
+};
+template <typename T, size_t N> struct get_elem_type<marray<T, N>> {
+  using type = T;
+};
+template <typename T, int N> struct get_elem_type<vec<T, N>> {
+  using type = T;
+};
+template <typename VecT, typename OperationLeftT, typename OperationRightT,
+          template <typename> class OperationCurrentT, int... Indexes>
+struct get_elem_type<SwizzleOp<VecT, OperationLeftT, OperationRightT,
+                               OperationCurrentT, Indexes...>> {
+  using type = typename get_elem_type<std::remove_cv_t<VecT>>::type;
+};
+
+template <typename T> using get_elem_type_t = typename get_elem_type<T>::type;
+
+#ifdef __FAST_MATH__
+template <typename T>
+struct use_fast_math
+    : std::is_same<std::remove_cv_t<get_elem_type_t<T>>, float> {};
+#else
+template <typename> struct use_fast_math : std::false_type {};
+#endif
+template <typename T> constexpr bool use_fast_math_v = use_fast_math<T>::value;
+
+template <typename> struct is_swizzle : std::false_type {};
+template <typename VecT, typename OperationLeftT, typename OperationRightT,
+          template <typename> class OperationCurrentT, int... Indexes>
+struct is_swizzle<SwizzleOp<VecT, OperationLeftT, OperationRightT,
+                            OperationCurrentT, Indexes...>> : std::true_type {};
+
+template <typename T> constexpr bool is_swizzle_v = is_swizzle<T>::value;
+
+template <typename T>
+constexpr bool is_vec_or_swizzle_v = is_vec_v<T> || is_swizzle_v<T>;
+
+// sycl::select(sgentype a, sgentype b, bool c) calls OpenCL built-in
+// select(sgentype a, sgentype b, igentype c). This type trait makes the
+// proper conversion for argument c from bool to igentype, based on sgentype
+// == T.
+template <typename T>
+using get_select_opencl_builtin_c_arg_type = typename std::conditional_t<
```

Comment:
This specific trait was from the existing builtins. I copied it verbatim to avoid misunderstanding the intention of the `long` cases and breaking the builtins using it.